src/System.Security.Cryptography.X509Certificates/src/Internal/Cryptography/Pal.Unix/UnixPkcs12Reader.cs

```diff
@@ -0,0 +1,670 @@
+                    }
+
+                    if (contentType == DecryptedSentinel)
+                    {
+                        ReadOnlyMemory<byte> content = rentedContents[0].Content;
+
+                        if (!MemoryMarshal.TryGetArray(content, out ArraySegment<byte> segment))
+                        {
+                            Debug.Fail("Couldn't unpack decrypted buffer.");
+                        }
+
+                        CryptoPool.Return(segment.Array, segment.Count);
+                        rentedContents[0].Content = default;
+                    }
+                }
+
+                ArrayPool<ContentInfoAsn>.Shared.Return(rentedContents, clearArray: true);
+            }
+
+            if (rentedCerts != null)
+            {
+                for (int i = _certCount - 1; i >= 0; --i)
+                {
+                    rentedCerts[i].Dispose();
+                }
+
+                ArrayPool<CertAndKey>.Shared.Return(rentedCerts, clearArray: true);
+            }
+        }
+
+        public void Decrypt(SafePasswordHandle password)
+        {
+            ReadOnlyMemory<byte> authSafeContents =
+                Helpers.DecodeOctetStringAsMemory(_pfxAsn.AuthSafe.Content);
+
+            bool hasRef = false;
+            password.DangerousAddRef(ref hasRef);
+
+            try
+            {
+                ReadOnlySpan<char> passwordChars = password.DangerousGetSpan();
+
+                if (_pfxAsn.MacData.HasValue)
+                {
+                    VerifyAndDecrypt(passwordChars, authSafeContents);
+                }
+                else if (passwordChars.IsEmpty)
+                {
+                    try
+                    {
+                        // Try the empty password first.
+                        // If anything goes wrong, try the null password.
+                        //
+                        // The same password has to work for the entirety of the file,
+                        // null and empty aren't interchangeable between parts.
+                        Decrypt("", authSafeContents);
+                    }
+                    catch (CryptographicException)
+                    {
+                        ContentInfoAsn[] partialSuccess = _safeContentsValues;
+                        _safeContentsValues = null;
+
+                        if (partialSuccess != null)
+                        {
+                            ArrayPool<ContentInfoAsn>.Shared.Return(partialSuccess);
+                        }
+
+                        Decrypt(null, authSafeContents);
+                    }
+                }
+                else
+                {
+                    Decrypt(passwordChars, authSafeContents);
+                }
+            }
+            catch (Exception e)
+            {
+                throw new CryptographicException(SR.Cryptography_Pfx_BadPassword, e);
+            }
+            finally
+            {
+                password.DangerousRelease();
```

Comment:
`if (hasRef)`?

Reply:
`hasRef` can't be false here.  DangerousAddRef throws if it fails (without modifying the ref input).  Since the call to DangerousAddRef isn't inside the try, this finally can only be reached if DangerousAddRef succeeded, therefore hasRef s true.

This pattern only works when there's only one DangerousAddRef, and it's the last thing before the try, but it's not uncommon.